doc/dev/crimson/pipelines.rst

```diff
@@ -0,0 +1,258 @@
+Because inherently a distributed storage system is distributed by its own nature, we can not assume
+that an OSD or the PGs hosted by it are always ready to serve a request targeting a shared state, not
+to mention the unpredictable hardware failures. But fortunately, even if an OSD is not able to serve
+a request right away, under most circumstances, it can postpone the request, prepare itself so that it
+can answer this request. It's quite common in daily life:
+
+  Great question, let me look into it and I will get back to you!
+
+Some questions involves multiple steps. To serve an I/O request from client, at
+least 7 stages are involved:
+
+.. ditaa::
+
+   +----+---------------------+
+   |cGRE| OSD wait for osdmap |
+   +----+----+----------------+
+             |
+             V
+   +----+-----------------+
+   |cGRE| OSD wait for PG |
+   +----+-----------------+
+             |
+             V
+   +----+--------------------+
+   |cBLU| PG wait for osdmap |
+   +----+--------------------+
+             |
+             V
+   +----+--------------------+
+   |cBLU| PG wait for active |
+   +----+--------------------+
+             |
+             V
+   +----+----------------------+
+   |cBLU| PG wait for recovery |
+   +----+----------------------+
+             |
+             V
+   +----+---------------------+
+   |cBLU| wait for object ctx |
+   +----+---------------------+
+             |
+             V
+   +----+------------------+
+   |cBLU| wait for process |
+   +----+------------------+
+
+             legend
+             +----------------+
+             |cGRE OSD stages |
+             +----------------+
+             |cBLU PG stage   |
+             +----------------+
+
+Two Requirements
+----------------
+
+An OSD client talks with each connected OSD over stateful connection. The client is allowed to
+perform asynchronous I/O with OSD in the sense that it can send another request without waiting
+for the completion of the previous one. But the contract of the RADOS protocol requires writes
+(and write ordered reads) performed on a single object to be performed in the order in which
+they were originally submitted, and only once. This behavior is implemented by both the client
+side and server side of librados.
+
+So far, we have two restrictions:
+
+#. sequential stages of a given request
+#. ordered execution of multiple write requests on a single object
+
+It's like visiting a history museum with friends.
+
+#. since the exhibition rooms in the museum are arranged in chronological order, it is advisable
+   to visit these rooms in the same order.
+#. before leaving the museum, visitors can leave a message on one of the notebooks provided by
+   the museum. But the visitors should do this in the same order in which they enter the museum.
+   For example, if both Joe and Jane decide to leave a message on the notebook with purple cover,
+   and Joe enters the museum before Jane, Jane cannot write on the notebook until Joe does.
```

Comment:
Ah, this is actually a good example, but it would be better to link it back to some piece of metadata in the actual OSD?

Reply:
added a table to explain the mapping.